### PR TITLE
Initial implementation

### LINKED ISSUE #1: Initial EDDBK Cart Module Implementation
## Description

The module will be responsible for the following:

- [x] When a booking is made via the wizard, a cart item is added for the booked service. The ID of the `in_cart` booking will be attached to this cart item.

- [x] When a cart item is deleted, the attached `in_cart` booking is deleted.

- [x] Validate the booking before checkout and show errors if applicable.

- [x] When a payment is processed, all attached `in_cart` bookings transition with `submit` and have their payment IDs updated.

## Related EDD hooks

* Added cart item: `edd_add_to_cart_item`

* Removed cart item `edd_pre_remove_from_cart`

* Validating cart items on checkout: `edd_checkout_error_checks`

* Payment Status update: `edd_update_payment_status`

### PR DESCRIPTION
**Pull request by mecha**
Resolves #1 

---

- [x] When a booking is made via the wizard, a cart item is added for the booked service. The ID of the `in_cart` booking will be attached to this cart item.

- [x] When a cart item is deleted, the attached `in_cart` booking is deleted.

- [x] When a payment is processed, all attached `in_cart` bookings transition with `submit` and have their payment IDs updated.

- [x] When EDD validates the cart items, a handler validates all bookings in the cart, by attempting a transition.

### CODE REVIEW

File: src/Module/RemoveBookingFromCartHandler.php [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+        try {
+            // Get the cart index from the event (provided by the EDD hook) and normalize it
+            $cartIndex = $event->getParam(0);
+            $cartIndex = $this->_normalizeInt($cartIndex);
+
+            // Get all the cart items - EDD does not provide a single cart item getter!
+            $cartItems = $this->eddCart->get_contents();
+            // Get the booking ID from the cart item at the index
+            $bookingId = $this->_containerGetPath($cartItems, [$cartIndex, $dataKey, $eddBkKey, $bookingIdKey]);
```
Comment (XedinUnknown) [inline src/Module/RemoveBookingFromCartHandler.php]: 😮 :weeping_joy:

File: src/Module/RemoveBookingFromCartHandler.php [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+            // Get the bookings that match - only 1 should match
+            $bookings = $this->bookingsSelectRm->select($c);
+            if ($this->_countIterable($bookings) !== 1) {
+                return;
+            }
+
+            // If the booking has a cart status, delete it
+            $booking = reset($bookings);
+            if ($booking->get('status') === Status::STATUS_IN_CART) {
+                $this->bookingsDeleteRm->delete($c);
```
Comment (XedinUnknown) [inline src/Module/RemoveBookingFromCartHandler.php]: I can see that, after jumping through hoops with EDD, you finally have the booking ID. Then you are actually removing the booking from the database after ascertaining that it exist and is in fact "in cart". So, why not abstract the database operations into a pehaps `_removeBookingFromCart()`, similarly to how you do it in the add-to-cart handler?

File: src/Module/SubmitBookingOnPaymentHandler.php [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        if ($oldStatus === 'publish' || $oldStatus === 'complete') {
+            return;
+        }
+        // Check if the payment is being updated to a status of "completion". If not, stop here.
+        if ($newStatus !== 'publish' && $newStatus !== 'complete') {
+            return;
+        }
+
+        // Get the payment meta
+        $paymentMeta = \edd_get_payment_meta($paymentId);
```
Comment (XedinUnknown) [inline src/Module/SubmitBookingOnPaymentHandler.php]: Global function use. Abstract to another method, or find a better way.

File: src/Module/SubmitBookingOnPaymentHandler.php [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+                }
+
+                // Get the booking
+                $_booking = reset($_bookings);
+                $_booking = $this->transitioner->transition($_booking, T::TRANSITION_SUBMIT);
+
+                // Prepare the change set
+                $_changeSet = [
+                    'payment_id' => $paymentId,
+                    'client_id'  => \edd_get_payment_customer_id($paymentId),
```
Comment (XedinUnknown) [inline src/Module/SubmitBookingOnPaymentHandler.php]: Global function use. Abstract to another method, or find a better way.

File: src/Module/SubmitBookingOnPaymentHandler.php [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+                    'status'     => $_booking->getStatus(),
+                ];
+
+                // Update booking
+                $this->bookingsUpdateRm->update($_changeSet, $_condition);
+            } catch (NotFoundExceptionInterface $exception) {
```
Comment (XedinUnknown) [inline src/Module/SubmitBookingOnPaymentHandler.php]: This is only supposed to be triggered `_containerGet()` on ln. 191. The catch should immediately follow that line, to protect against runaway not-found-exceptions being accidentally thrown by other calls in the `try`.

File: src/Module/ValidateCartBookingHandler.php [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+            $bookings  = $this->bookingsSelectRm->select(
+                $b->eq(
+                    $b->ef('booking', 'id'),
+                    $b->lit($bookingId)
+                )
+            );
+
+            // Check that only 1 booking was retrieved
+            if ($this->_countIterable($bookings) !== 1) {
+                \edd_set_error('eddbk_invalid_booking_id', $this->__('A cart item has an invalid booking ID.'));
```
Comment (XedinUnknown) [inline src/Module/ValidateCartBookingHandler.php]: Global function use. Abstract to another method, or find a better way.

File: src/Module/ValidateCartBookingHandler.php [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+            // Get the booking
+            $booking = reset($bookings);
+
+            try {
+                // Validate it
+                $this->_validateBooking($booking);
+            } catch (ValidationFailedExceptionInterface $exception) {
+                // Register all validation errors as EDD checkout errors
+                foreach ($exception->getValidationErrors() as $key => $error) {
+                    \edd_set_error(
```
Comment (XedinUnknown) [inline src/Module/ValidateCartBookingHandler.php]: Tired of typing. Guess what the problem is.

File: src/Module/ValidateCartBookingHandler.php [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    /**
+     * Retrieves the validation failure exception from an exception's previous exceptions chain.
+     *
+     * @since [*next-version*]
+     *
+     * @param RootException $exception The exception to search.
+     *
+     * @return ValidationFailedExceptionInterface|null The found validation failure exception or null if not found.
+     */
+    protected function _getValidationFailedException(RootException $exception)
```
Comment (XedinUnknown) [inline src/Module/ValidateCartBookingHandler.php]: "resolve" pls.

**Review by XedinUnknown — CHANGES_REQUESTED** [review]
Overall, a pleasure to read. But some small but crucial improvements here and there.

File: src/Module/ValidateCartBookingHandler.php (hunk 7 above)
Comment (mecha) [reply src/Module/ValidateCartBookingHandler.php]: It's on line 196? I'll add a new line so that it will be on 197 😆 🤣

**Review by XedinUnknown — APPROVED** [review]

**Pull request merged**